lib/minfraud/transaction.rb

```diff
@@ -28,60 +32,135 @@ class Transaction
+      :requested_type
+    ].freeze
+
+    INPUT_NUMERIC_ATTRIBUTES = [:amount].freeze
+
+    REQUIRED_ATTRIBUTES = [:ip, :txn_id, :license_key].freeze
+
+    CONVERT_STRING_ATTRIBUTES = [:shop_id].freeze
+
+    # Initializes the Transaction using parameters set from the block
+    # @raise [TransactionError] if parameters are set incorrectly
     def initialize
       yield self
       unless has_required_attributes?
         raise TransactionError, 'You did not set all the required transaction attributes.'
       end
       validate_attributes
+      freeze
     end
 
     # Hash of attributes that have been set
     # @return [Hash] present attributes
     def attributes
-      attrs = [:ip, :city, :state, :postal, :country, :license_key, :ship_addr, :ship_city, :ship_state, :ship_postal, :ship_country, :email_domain, :email_md5, :phone, :bin, :session_id, :user_agent, :accept_language, :txn_id, :amount, :currency, :txn_type, :avs_result, :cvv_result, :requested_type, :forwarded_ip, :shop_id]
-      attrs.map! do |a|
-        [a, send(a)]
-      end
-      Hash[attrs]
+      Hash[ATTRIBUTES.map { |a| [a, send(a)] }].compact
     end
 
     private
 
     # Ensures the required attributes are present
     # @return [Boolean]
     def has_required_attributes?
-      ip && txn_id && license_key
+      REQUIRED_ATTRIBUTES.none? { |attr| send(attr).nil? || send(attr).empty? }
     end
 
     # Validates the types of the attributes
-    # @return [nil, TransactionError]
+    # @raise [TransactionError] if present attributes are not valid
+    # @return [nil]
     def validate_attributes
-      [:ip, :city, :state, :postal, :country, :txn_id, :license_key].each { |s| validate_string(s) }
+      INPUT_STRING_ATTRIBUTES.each { |attr| validate_string(attr) }
+      INPUT_NUMERIC_ATTRIBUTES.each { |attr| validate_number(attr) }
+      CONVERT_STRING_ATTRIBUTES.each { |attr| convert_to_string(attr) }
+      validate_cvv
     end
 
     # Given the symbol of an attribute that should be a string,
     # it checks the attribute's type and throws an error if it's not a string.
     # @param attr_name [Symbol] name of the attribute to validate
-    # @return [nil, TransactionError]
+    # @raise [TransactionError] if attribute is not a string
+    # @return [nil]
     def validate_string(attr_name)
-      attribute = self.send(attr_name)
+      attribute = send(attr_name)
       if attribute && !attribute.instance_of?(String)
         raise TransactionError, "Transaction.#{attr_name} must be a string"
       end
     end
 
+    # Given the symbol of an attribute that should be a number,
+    # it checks the attribute's type and throws an error if it's not a number.
+    # @param attr_name [Symbol] name of the attribute to validate
+    # @raise [TransactionError] if attribute is not a number
+    # @return [nil]
+    def validate_number(attr_name)
+      attribute = send(attr_name)
+      if attribute && !attribute.is_a?(Numeric)
```

Comment:
Here the Numeric type is used as an all encompassing type of all attributes that require the validation of a number.      

The actual type of `amount` attribute is `BigDecimal` and `BigDecimal` inherits from Numeric class.

Reply:
Yep! Should this validation ensure that a `BigDecimal` is needed? Other users of this client may pass in something else, so I went with `Numeric`. (I also included a test to make sure `BigDecimal` is properly accepted)